src/operators/mod.rs

```diff
@@ -89,138 +100,142 @@ pub(crate) fn inspect_epoch(inspector: &TdPyCallable, epoch: &u64, item: &TdPyAn
-            Some(acc) => with_traceback!(py, reducer.call1(py, (acc.clone_ref(py), value))).into(),
-            None => value,
-        };
-        *acc = Some(updated_acc);
-        debug!(
-            "{}, reducer:{:?}, key:{:?}, updated_acc:{:?}",
-            log_func!(),
-            reducer,
-            _key,
-            acc
-        );
-    });
-}
-
-pub(crate) fn reduce_epoch_local(
-    reducer: &TdPyCallable,
-    accs: &mut HashMap<StateKey, TdPyAny>,
-    all_key_value_in_epoch: &Vec<(StateKey, TdPyAny)>,
-) {
-    Python::with_gil(|py| {
-        let _current = thread::current();
-        let id = _current.id();
-        for (key, value) in all_key_value_in_epoch {
-            let acc = accs.entry(key.clone());
-            debug!(
-                "thread:{:?}, {}, reducer:{:?}, key:{:?}, value:{:?}, acc:{:?}",
-                id,
-                log_func!(),
-                reducer,
-                key,
-                value,
-                acc
-            );
-            acc.and_modify(|acc| {
-                *acc = with_traceback!(py, reducer.call1(py, (acc.clone_ref(py), value))).into();
-                debug!(
-                    "{}, reducer:{:?}, key:{:?}, value:{:?}, updated_acc:{:?}",
-                    log_func!(),
-                    reducer,
-                    key,
-                    value,
-                    *acc
-                );
-            })
-            .or_insert(value.clone_ref(py));
-        }
-    });
+    if let Poll::Ready(Some(value)) = next_value {
+        Python::with_gil(|py| {
+            let updated_acc: TdPyAny = match acc {
+                // If there's no previous state for this key, use the
+                // current value.
+                None => value,
+                Some(acc) => {
+                    let updated_acc = with_traceback!(
+                        py,
+                        reducer.call1(py, (acc.clone_ref(py), value.clone_ref(py)))
+                    )
+                    .into();
+                    debug!("reduce for key={key:?}: reducer={reducer:?}(acc={acc:?}, value={value:?}) -> updated_acc={updated_acc:?}",);
+                    updated_acc
+                }
+            };
+
+            let should_emit_and_discard_acc: bool = with_traceback!(py, {
+                let should_emit_and_discard_acc_pybool: TdPyAny =
+                    is_complete.call1(py, (updated_acc.clone_ref(py),))?.into();
+                should_emit_and_discard_acc_pybool.extract(py).map_err(|_err| PyTypeError::new_err(format!("return value of `is_complete` in reduce operator must be a bool; got `{should_emit_and_discard_acc_pybool:?}` instead")))
+            });
+            debug!("reduce for key={key:?}: is_complete={is_complete:?}(updated_acc={updated_acc:?}) -> should_emit_and_discard_acc={should_emit_and_discard_acc:?}");
+
+            if should_emit_and_discard_acc {
+                (None, Some(updated_acc), None)
```

Comment:
I think it would probably be awkward to wrap this in another type but it's tricky to read the 3 tuple of options 🤔

Reply:
Not a bad idea. It's a little verbose, but we can try it out.